pygmt/clib/conversion.py

```diff
@@ -304,7 +306,12 @@ def strings_to_ctypes_array(strings: Sequence[str]) -> ctp.Array:
     >>> [s.decode() for s in ctypes_array]
     ['first', 'second', 'third']
     """
-    return (ctp.c_char_p * len(strings))(*[s.encode() for s in strings])
+    try:
+        bytes_string_list = [s.encode() for s in strings]
+    except AttributeError:  # 'pyarrow.StringScalar' object has no attribute 'encode'
+        # Convert pyarrow.StringArray to Python list first
+        bytes_string_list = [s.encode() for s in strings.to_pylist()]  # type: ignore[union-attr]
+    return (ctp.c_char_p * len(strings))(*bytes_string_list)
```

Comment:
What about using `np.asarray` to convert `strings` into a numpy array first?
```suggestion
    return (ctp.c_char_p * len(strings))(*[s.encode() for s in np.asarray(strings)])
```

Here are benchmarks:
```python
In [1]: import pyarrow as pa

In [2]: import numpy as np

In [3]: stringlist = ["hello", "world", "test"]

In [4]: stringarray = pa.array(stringlist)

In [5]: %timeit [s.encode() for s in stringlist]
196 ns ± 0.879 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [6]: %timeit [s.encode() for s in stringarray.to_pylist()]
2.56 μs ± 14.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [7]: %timeit [s.encode() for s in np.asarray(stringlist)]
2.13 μs ± 22.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [8]: %timeit [s.encode() for s in np.asarray(stringarray)]
3.43 μs ± 19.9 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```
For a much larger string array:
```python
In [19]: stringlist = ["hello", "world", "test"] * 1000

In [20]: stringarray = pa.array(stringlist)

In [21]: %timeit [s.encode() for s in stringlist]
115 μs ± 484 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [22]: %timeit [s.encode() for s in stringarray.to_pylist()]
2.17 ms ± 17.6 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [23]: %timeit [s.encode() for s in np.asarray(stringlist)]
1.07 ms ± 4.88 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [24]: %timeit [s.encode() for s in np.asarray(stringarray)]
226 μs ± 1.26 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```
- **Pros:** consistent way to handle a sequence, a numpy array and pyarrow.StringDtype.
- **Cons**: 10x slower for string list, but I guess it's OK since it's very fast even for a large array

Also need to note that, currently `strings_to_ctypes_array` is only used in four functions, `put_vector`/`put_strings`/`virtualfile_from_stringio`/`call_module`.  `put_vector` and `put_string` are called in `virtualfile_from_vectors` only and before calling these two functions, the *array* has been converted to a numpy array by the `vectors_to_arrays` function at https://github.com/GenericMappingTools/pygmt/blob/96b12634d95a0fd2a262bcf5245a942cb4077e69/pygmt/clib/session.py#L1384.

So the performance reduction should be negligible:
```
In [1]: import numpy as np

In [2]: stringlist = ["hello", "world", "test"] * 1000

In [3]: stringarray = np.asarray(stringlist)

In [4]: %timeit [s.encode() for s in stringarray]
671 μs ± 13.9 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [5]: %timeit [s.encode() for s in np.asarray(stringarray)]
688 μs ± 3.3 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

Reply:
Thanks for the benchmarks. I re-ran them on my computer (plus tried `np.char.encode` mentioned in https://github.com/GenericMappingTools/pygmt/pull/3137#issue-2203772747), and think we should go with `[s.encode() for s in np.asarray(strings)]` as you suggest:

```python
%timeit [s.encode() for s in stringlist]
65.1 μs ± 579 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

%timeit [s.encode() for s in stringarray.to_pylist()]
1.35 ms ± 5.1 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

%timeit [s.encode() for s in np.asarray(stringlist)]
588 μs ± 6.61 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

%timeit [s.encode() for s in np.asarray(stringarray)]
129 μs ± 599 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

%timeit np.char.encode(stringlist)
784 μs ± 7.09 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

%timeit np.char.encode(np.asarray(stringarray, dtype=str))
841 μs ± 11 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```